src/tree_types.ml

```diff
@@ -0,0 +1,265 @@
+      ; c_ari : int
+      (** Arity of the considered constructor.  A same symbol representation
+          may generate several constructors with different arities.  For
+          instance, with [g: N ⇒ N ⇒ N], [rule f g (g &x) → r] uses [g] with
+          arity 1 and 2. *) }
+  | Abst
+  (** An abstraction*)
+  | Vari of string
+  (** A bound variable with a name. *)
+
+(** [tc_pp o c] prints tree constructor [c] to output channel [o]. *)
+let tc_pp : treecons pp = fun oc -> function
+  | Abst    -> Format.fprintf oc "λ"
+  | Vari(s) -> Format.pp_print_string oc s
+  | Symb(t) -> Format.fprintf oc "%s<:%d" t.c_sym t.c_ari
+
+(** [tc_compare c d] is a comparison function for constructors; more efficient
+    than the pervasive. *)
+let tc_compare : treecons -> treecons -> int = fun ca cb ->
+  match ca, cb with
+  | Symb(a), Symb(b) ->
+      begin match Int.compare a.c_ari b.c_ari with
+      | 0 ->
+          begin match String.compare a.c_sym b.c_sym with
+          | 0 -> Pervasives.compare a.c_mod b.c_mod
+          | x -> x
+          end
+      | x -> x
+      end
+  | x, y                 -> Pervasives.compare x y
+
+(** A simple comparable type. *)
+module type Comparable =
+sig
+
+  (** Type of data. *)
+  type t
+
+  (** A mapping is considered {i big} if the number of bindings exceeds the
+      threshold.  The threshold [t] should be such that
+      + a {!constructor:Light} mapping with [t - 1] elements should have an
+        access time smaller than a {!constructor:Heavy} one,
+      + a {!constructor:Heavy} mapping with [t + 1] elements should have an
+        access time smaller than a {!constructor:Light} one. *)
+  val threshold : int
+
+  (** [compare x y] is [< 0] if [x < y], [= 0] if [x = y] and [> 0]
+      otherwise. *)
+  val compare : t -> t -> int
+end
+
+(** Subset of a mapping with {!type:constructor} as keys. *)
+module type MapSig =
+sig
+  (** Type of keys. *)
+  type key
+
+  (** Type of a mapping. *)
+  type +'a t
+
+  (** The empty map. *)
+  val empty : 'a t
+
+  (** [is_empty m] returns whether [m] is empty; more efficient than comparing
+      to [empty] (avoids a conversion). *)
+  val is_empty : 'a t -> bool
+
+  (** [add k e m] adds element [e] with key [k] to mapping [m]. *)
+  val add : key -> 'a -> 'a t -> 'a t
+
+  (** [bindings m] returns an associative list [(k, e)] with [k] key and [e]
+      element. *)
+  val bindings : 'a t -> (key * 'a) list
+
+  (** [find k m] returns the element mapped by [k]
+      through [m].
+      @raise Not_found if [k] is not a key of [m]. *)
+  val find : key -> 'a t -> 'a
+
+  (** [iter f m] iters function [f] on mapping [m]. *)
+  val iter : (key -> 'a -> unit) -> 'a t -> unit
+
+  (** [map f m] maps [f] on mapping m. *)
+  val map : ('a -> 'b) -> 'a t -> 'b t
+end
+
+(** [MkMap(C)] creates a mapping from values of [C].  Very small mappings are
+    treated differently to have faster access. *)
+module MkMap(C:Comparable) : (MapSig with type key = C.t) =
```

Comment:
Is this module really worth it? I'm not sure that this buys us anything in practice. Did you bench that?
If not, I would simplify all that using
```ocaml
module TCMap = Map.Make(TreeCons)
```

Reply:
No I don't think it is relevant, I'll remove it indeed